### ISSUE TITLE
Display version of add-on save is from when loading a savefile from a UMC campaign

### ISSUE THREAD

**Issue by cooljeanius** [wesnoth/wesnoth#6150]
So, recently I was trying to view some old replays from "A New Order" for the purposes of developing it; however, replays from old versions of the campaign throw OOS on me due to a changed unit. When loading a savefile or replay, the "load save" dialog includes the BfW version, but that's the only version it includes:
<img width="255" alt="Screen Shot 2021-10-02 at 18 31 50" src="https://user-images.githubusercontent.com/1388660/135763707-6eb5f966-7e1d-437f-a36f-6667bfb8d5b9.png">
Also displaying the version number of the add-on a scenario is from could be helpful for add-on authors in determining which old saves to bother loading or not when trying to debug something.

**Comment by Wedge009** [comment]
Just wondering: is version numbering actually enforced for add-ons? I mean, it's not ever left blank.

As an aside, there was someone in #wesnoth-dev bemoaning having to change version numbers for their add-on with every upload to 1.15 server (apparently 1.14 doesn't have this restriction). If some authors aren't properly maintaining their add-on versions, it wouldn't make this update very useful. On the other hand, I suppose that's really a fault of the particular add-on, not the feature.

**Comment by cooljeanius** [comment]
> Just wondering: is version numbering actually enforced for add-ons? I mean, is it possible to leave it blank, for example?
> 
> As an aside, there was someone in #wesnoth-dev bemoaning having to change version numbers for their add-on with every upload to 1.15 server (apparently 1.14 doesn't have this restriction). If some authors aren't properly maintaining their add-on versions, it wouldn't make this update very useful. On the other hand, I suppose that's really a fault of the particular add-on, not the feature.

Yeah I saw that user's complaints and am glad that everyone explained the importance of versioning in response. I would support changes to make Wesnoth even stricter about version numbers, including possibly enforcing the Semantic Versioning standard. Although, there's one alternative possibility I considered: instead of getting stricter about version numbers, instead require that all add-ons be stored in git repositories, and then just use commit hashes instead of version numbers. In which case this bug would be to display the commit hash instead.

**Comment by Wedge009** [comment]
At the risk of making this discussion tangential to the actual issue request, I just wonder about the suggestion to make add-on authors use git (which is also related to #1458). Aside from potentially raising the bar for developing UMC, from an 'average user' perspective it isn't immediately obvious how old a version might be just looking at the commit hash. At least semantic version convention attempts to convey some sense consistency with version numbers that's immediately obvious to users.

**Comment by cooljeanius** [comment]
> I was looking at an ANO replay @cooljeanius sent me and found that the addon version is stored in the replay metadata. This can simply be read and displayed for implementing this suggestion.
> 
> ```
> 	[addon]
> 		id="A_New_Order"
> 		name="A New Order"
> 		version="1.2.19"
> 	[/addon]
> ```

Just how "simply" are we talking here? Like, would this be material for the "good first issue" label?

**Comment by cooljeanius** [comment]
> Like, would this be material for the "good first issue" label?

Eh, I guess I'll just add the label anyways, and if someone disagrees, they can remove it...

**Comment by Wedge009** [comment]
A lot of the 'good first issue' aren't that great for newcomers anyway. But that aside, I suppose I didn't really consider this previously because I normally don't play add-on campaigns. But aside from testing candidates for main-line I realise there's also An Orcish Incursion that was booted from main-line in 1.16.

I wonder: why are the `[addon]` entries grouped under `[multiplayer]`? It's a single-player save file. I'm not really familiar with the save format anyway.

I had a look at this and I really don't think it should have the 'good first issue' label. Conceptually it's straightforward but the implementation details look tricky to me. I think only the top-level attributes in the save file are loaded via `game.summary()` (`game_load::display_savegame_internal()` in `game_load.cpp`) - I think to get access to children info of the save file root one has to call `read_save_file()` to read the entire contents. Either that, or change the save file format to include the campaign add-on version in the summary.

Unless I got this save file handling wrong and it turns out to be really simple... I was focusing on these functions:
https://github.com/wesnoth/wesnoth/blob/2630b5bb630fd20cdc3f9fe98292781381798178/src/gui/dialogs/game_load.cpp#L183-L186
https://github.com/wesnoth/wesnoth/blob/2630b5bb630fd20cdc3f9fe98292781381798178/src/gui/dialogs/game_load.cpp#L241-L243
https://github.com/wesnoth/wesnoth/blob/2630b5bb630fd20cdc3f9fe98292781381798178/src/gui/dialogs/game_load.cpp#L456-L458

**Comment by cooljeanius** [comment]
> I had a look at this and I really don't think it should have the 'good first issue' label. Conceptually it's straightforward but the implementation details look tricky to me.

OK, I'll remove it again.
